src/jit/codegenlinear.cpp

```diff
@@ -2304,13 +2304,26 @@ void CodeGen::genCodeForJumpTrue(GenTreeOp* jtrue)
     assert(compiler->compCurBB->bbJumpKind == BBJ_COND);
     assert(jtrue->OperIs(GT_JTRUE));
 
-    GenCondition condition = GenCondition::FromRelop(jtrue->gtGetOp1());
+    GenTreeOp*   relop     = jtrue->gtGetOp1()->AsOp();
+    GenCondition condition = GenCondition::FromRelop(relop);
 
     if (condition.PreferSwap())
     {
         condition = GenCondition::Swap(condition);
     }
 
+#if defined(_TARGET_XARCH_)
+    if ((condition.GetCode() == GenCondition::FNEU) && GenTree::Compare(relop->gtGetOp1(), relop->gtGetOp2(), true))
```

Comment:
I'd stay away from `GenTree::Compare`. It's expensive, potentially incorrect and unlikely to be very helpful. The check should test only local variables.

Reply:
Hmmm you are probably right. If so, we should eliminate this and/or replace it with something that is more accurate and useful.

Reply:
> The check should test only local variables.

Ah, this is codegen, I've done this in the past but in lowering. In codegen this can probably check if the 2 operands are in the same register. I would expect that if both operands use the same variable they'll get the same register. @CarolEidt ?

Reply:
>  would expect that if both operands use the same variable they'll get the same register.

I was actually seeing both be the same local in the samples I was checking (although I was also not testing particularly interesting samples; basically just calling `IsNaN(x)` and `if (x != x)` on a non-inlinable method where `x` was a method parameter)

Reply:
@CarolEidt, this actually appears to be fairly common in the samples I am checking:
![image](https://user-images.githubusercontent.com/10487869/67051800-06993800-f0f1-11e9-866a-fbd71c901da8.png)

Reply:
I am also seeing codegen where both registers are the same, but the above doesn't seem uncommon.

Reply:
What's `src` and `dst`? If `src` has reg num = REG_COUNT that probably means that it is contained or something like that.

Reply:
Ah, I see. I forgot to disable tiered compilation for this scenario, forcing tier 1 shows that the right thing happens and that just checking the reg num gets the full diff for S.P.Corelib.

In this case, this was for (specifically for `x != x`):
```csharp
[MethodImpl(MethodImplOptions.NoInlining)]
static void Test(float x)
{
    Console.WriteLine(float.IsNaN(x));

    if (x != x)
    {
        Console.WriteLine("NaN");
    }
    else
    {
        Console.WriteLine(x);
    }
}
```